pkg/kubelet/status/status_manager.go

```diff
@@ -868,9 +875,29 @@ func mergePodStatus(oldPodStatus, newPodStatus v1.PodStatus, couldHaveRunningCon
 			podConditions = append(podConditions, c)
 		}
 	}
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.PodDisruptionConditions) {
+		// update pod disruption conditions if a new one is present, otherwise preserve the old one
+		cType := v1.AlphaNoCompatGuaranteeDisruptionTarget
+		_, condition := podutil.GetPodConditionFromList(newPodStatus.Conditions, cType)
+		if condition == nil {
+			_, condition = podutil.GetPodConditionFromList(oldPodStatus.Conditions, cType)
+		}
+		if condition != nil {
+			podConditions = append(podConditions, *condition)
+		}
+	}
+
 	for _, c := range newPodStatus.Conditions {
 		if kubetypes.PodConditionByKubelet(c.Type) {
-			podConditions = append(podConditions, c)
+			if utilfeature.DefaultFeatureGate.Enabled(features.PodDisruptionConditions) {
+				// skip copying of pod disruption conditions here, as they are copied (if any present) already
+				if c.Type != v1.AlphaNoCompatGuaranteeDisruptionTarget {
+					podConditions = append(podConditions, c)
```

Comment:
one concern i have over doing this:  the pod status itself (i.e. phase, reason, etc) is owned by kubelet, and kubelet will always update the status with the latest one (ignoring status on server). However, for this condition, if it existed in the old pod status, we will always copy it over. Is it possible we will need up in a situation where the pod status and this conditions will diverge as a result? For example:

1. Pod created
2. Pod killed (evicted, etc)
3. `AlphaNoCompatGuaranteeDisruptionTarget` Condition set
4. Pod restarts (say it has some restart policy)
5. Pod is now running, `phase=Running`, but has the still has the `AlphaNoCompatGuaranteeDisruptionTarget` condition (`TerminatedByKubelet`) still attached to it

That doesn't sound like something we want -- i.e. `TerminatedByKubelet` condition should only be added for terminal pods (i.e. `phase=Failed/Succeeded`)

I suppose for now, we only add the condition in places where the pod is placed in terminal phase (`Failed`) (shutdown manager and eviction manager) so it can't start restart again. But I'm thinking this may change in the future if for example we add memory OOM condition?

Maybe we should only append this condition if this pod status has a terminal phase to ensure the scenario above will not happen?

Reply:
Indeed we don't add it in scenarios when the pod can restart. However, the pod still can get the condition when the phase is `Running` as the pod may still have containers running with grace period: https://github.com/kubernetes/kubernetes/blob/548d135ea6b3b9a68d31c981b65e1f9e9643ea58/pkg/kubelet/status/status_manager.go#L888. 

This is a problem we solved already in case of preemption. This is the solution:
- we call the condition `DisruptionTarget` as the actual disruption may not happen if scheduler fails to delete the pod, but succeeds with the patch for adding the condition,
- job controller failure policy ignores pods until they are in the `Failed` phase (so the `DisruptionTarget` condition remains transparent until then)
- stale `DisruptionTarget` conditions (added > 2min ago to a pod without `deletionTimestamp`) are cleaned up by a periodic disruption worker

Reply:
Still, I'm wondering that there could an issue here. If the pod grace period is long (>2min), then pod disruption controller could clean up the condition (set its status to `False`) before the actual transition into the `Failed` state. I'm wondering, maybe we should exclude conditions added by kubelet from being processed by the disruption worker?

Reply:
After discussing with @alculquicondor we decided to solve the issue by excluding the `DisruptionTarget` conditions added by kubelet from handling by disruption controller, based on the `reason` field being equal to `TerminationByKubelet`.